### PR TITLE
Doe categorical second attempt

### PR DESCRIPTION
**Pull request by ufukguenes**
Reopens #243

### CODE REVIEW

**Comment by ufukguenes** [comment]
@jduerholt regarding pyright.
As far as I can tell pyright has two problems with the code. First, when using the get_features method the returned type is of type Features which does not guarantee that I can use them like a RelaxableBinaryInput, however I can be sure that they are only RelaxableBinaryInputs as that's what I am asking for. Maybe I should add a type ignore to these instances. A cleaner option would be to make the return type of get_features generic. 
The second problem is, that the NChooseKConstraint is allowed to have Inputs which are not ContinousInputs or DiscreteInputs which might lead to a problem, but this this also not taken in to account with the previous implementation. So here as well adding type ignore could be a solution

**Comment by jduerholt** [comment]
Ok, then just `type: ignore` them.

File: bofire/strategies/doe_strategy.py [hunk 1]
```diff
@@ -20,22 +37,139 @@ def __init__(
...
+                n_experiments=_candidate_count,
+                fixed_experiments=self.candidates,
+                verbose=self.data_model.verbose,
+                partially_fixed_experiments=self._partially_fixed_experiments_for_next_design,
+                categorical_groups=all_new_categories,
+            )
+        elif self.data_model.optimization_strategy in [
+            "branch-and-bound",
+            "default",
+            "partially-random",
```
Comment (dlinzner-bcs) [inline bofire/strategies/doe_strategy.py:153]: why do we have again all these strategies that map to the same optimizer?

Comment (ufukguenes) [reply bofire/strategies/doe_strategy.py:153]: depending on which Inputs are in the domain, need to use the BaB regardless of which strategy is provided. For example, if there are CategoricalInputs and NChooseKConstraints and the strategy is set to partially-random. We need solve the NChosseKConstraints by using the already existing strategy where we randomly sample solution, but the CategoricalInputs have to be solved with the BaB.

Comment (dlinzner-bcs) [reply bofire/strategies/doe_strategy.py:153]: Ok.  Can you please add comments that explain these cases and what the strategies mean? And maybe "partially-random" is not very idiomatic. It is referring to  the current NChooseKConstraint implementation, correct? Maybe call it "random-nchoosek-realization"?

File: bofire/strategies/doe/utils_features.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+from bofire.data_models.features.continuous import ContinuousInput
+
+
+class RelaxableBinaryInput(ContinuousInput):
+    """Base class for all binary input features. It behaves like a continuous inputs to allow for easy relaxation.
+    It is not a true binary variable as it does not allow the variable to be either 0 or 1 while prohibiting all values
+    in between.
+
+    """
+
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply bofire/strategies/doe/utils_features.py:18]: bounds: Tuple(float, float) = (0,1)

File: bofire/strategies/doe/utils_features.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    """Feature with discrete ordinal values allowed in the optimization.
+
+    Attributes:
+        key(str): key of the feature.
+        values(List[float]): the allowed discrete values during the optimization.
+    """
+
+    values: List[float]
+
+    def __init__(self, **kwargs):
```
Comment (jduerholt) [inline bofire/strategies/doe/utils_features.py:69]: not needed, do it in the above.

File: bofire/strategies/doe/utils_features.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+from typing import List, Tuple
+
+import numpy as np
+import pandas as pd
+from pydantic import root_validator
+
+from bofire.data_models.features.continuous import ContinuousInput
+
+
+class RelaxableBinaryInput(ContinuousInput):
```
Comment (jduerholt) [inline bofire/strategies/doe/utils_features.py:12]:
```suggestion
class _RelaxableBinaryInput(ContinuousInput):
```

**Comment by ufukguenes** [comment]
@jduerholt @dlinzner-bcs as discussed, I removed the two features I added and use a Dict/ List to keep track of the relaxed variables

**Comment by jduerholt** [comment]
> @jduerholt @dlinzner-bcs as discussed, I removed the two features I added and use a Dict/ List to keep track of the relaxed variables

Great. Thank you very much!

File: bofire/strategies/doe_strategy.py [hunk 5]
```diff
@@ -20,22 +32,138 @@ def __init__(
     ):
         super().__init__(data_model=data_model, **kwargs)
         self.formula = data_model.formula
+        self.data_model = data_model
+        self._partially_fixed_experiments_for_next_design = None
+
+    def tell(
```
Comment (jduerholt) [inline bofire/strategies/doe_strategy.py:38]: Hmm, for what is this `tell` needed?

File: tests/bofire/data_models/specs/strategies.py [hunk 6]
```diff
@@ -125,6 +125,8 @@
     lambda: {
         "domain": domain.valid().obj().dict(),
         "formula": "linear",
+        "optimization_strategy": "default",
+        "verbose": False,
```
Comment (jduerholt) [inline tests/bofire/data_models/specs/strategies.py:129]: What is this ´verbose´ flag doing?

File: tutorials/basic_examples/Reaction_Optimization_Example.ipynb
Comment (jduerholt) [inline tutorials/basic_examples/Reaction_Optimization_Example.ipynb:1]: Is this indended that so much is deleted from all these notebooks? No, or? It looks like, that all the output is gone. Have you modified them? This relates to all the notebooks ...

File: bofire/strategies/doe_strategy.py (hunk 5 above)
Comment (jduerholt) [reply bofire/strategies/doe_strategy.py:38]: Especially as you overwrite `tell` and not just `_tell`. That's looks a bit dangerous ...

File: bofire/strategies/doe_strategy.py [hunk 7]
```diff
@@ -20,22 +32,138 @@ def __init__(
...
+                raise ValueError(
+                    f"No variance in experiments for fixed feature {feature.key}"
+                )
+        self._tell()
+
+    def _tell(self) -> None:
+        self.set_candidates(
+            self.experiments[self.domain.get_feature_keys(includes=Input)]
+        )
+        raise NotImplementedError(
```
Comment (jduerholt) [inline bofire/strategies/doe_strategy.py:78]: Also this is a bit cumbersome, the idea is that one has a generic `tell` method in `Strategy`, which is then calling a strategy dependent `_tell`. In principal, I do not see why a tell is needed at all, we have already the pending candidates to take previous experiments into account ...

**Review by jduerholt — CHANGES_REQUESTED** [review]
Thank you very much. Looks much better know. I let some small comments. We should especially have a closer look what happens to the tutorial notebooks.

File: bofire/strategies/doe_strategy.py (hunk 7 above)
Comment (dlinzner-bcs) [reply bofire/strategies/doe_strategy.py:78]: Whoops. I told Ufuk to implement tell. My bad. @ufukguenes let us use set_candidates() then instead.

File: tests/bofire/data_models/specs/strategies.py (hunk 6 above)
Comment (ufukguenes) [reply tests/bofire/data_models/specs/strategies.py:129]: If true, the method prints out information about the optimization process. So for BaB, the number of branches explored and the current lower bound.

**Comment by ufukguenes** [comment]
@jduerholt I converted the tell method to set_candidates. I had to over wright the method in the child class as we want to support partially defined experiments that can be filled up by the optimizer. We think that is a valuable feature by it self. The parent set_candidates checked for None values which we use to define the partially defined experiments.

**Comment by jduerholt** [comment]
> @jduerholt I converted the tell method to set_candidates. I had to over wright the method in the child class as we want to support partially defined experiments that can be filled up by the optimizer. We think that is a valuable feature by it self. The parent set_candidates checked for None values which we use to define the partially defined experiments.

This is a very smart idea. In a seperate PR, we should update the base class to make it a general feature.

**Review by jduerholt — APPROVED** [review]
@ufukguenes: thanks for all your efforts!

@ufukguenes  @dlinzner-bcs : Feel free to merge it ;)

**Pull request merged**